### DISCUSSION TITLE
wiremock-spring-boot with a BDD Cucumber step definition

### DISCUSSION THREAD

**Discussion by corstad** [maciejwalkowiak/wiremock-spring-boot#22] [Q&A]
I have successfully used this library for a @SpringBootTest. I am now trying to utilize it on a Spring Boot Cucumber test and am having issues. The @InjectWireMock isn't being wired for the test.

Has anyone been able to use this library with a Spring Boot Cucumber test?

**Comment by maciejwalkowiak** [comment]
Can you publish somewhere a sample that does not work?

**Comment by McKratt** [comment]
@maciejwalkowiak In cucumber with Spring Boot you have a class which define the runtime context for Cucumber Test. It is on this class you put the annotation `@EnableWireMock`. Like following : 
```java
@SpringBootTest(webEnvironment = RANDOM_PORT)
@CucumberContextConfiguration
@EnableWireMock({
        @ConfigureWireMock(name = "my-service", property = "application.url"),
        @ConfigureWireMock(name = "positions-service", property = "position.url"),
        @ConfigureWireMock(name = "crm-service", property = "crm.url")
})
@Testcontainers
public class BootstrapCucumberContextConfiguration {
```
After that I try to inject defined Wiremock server into Step class as we do for Spring Bean : 
```java 
public class Steps  {
    @Autowired
    private BeanInterface bean;
    @InjectWireMock("my-service")
    private WireMockServer myService;

    @Given("my step...")
    public void step1() {
        myService.stubFor(...);
    }
```
When you run this configuration you end up with a null myService which provoque a NullPointerException (of course).

**Comment by McKratt** [comment]
In addition, Cucumber use the suite Junit 5 feature through the annotation `@Suite` and its own engine. From my minimalist comprehension of Junit5. Running a TestSuite with a custom engine doesn't seem to "launch" extensions. So the `WireMockSpringExtension` can not inject wiremock server in classes in the test context.

**Reply by McKratt** [reply]
Confirmed ! Custom Engine doesn't use Extension.  Only the Jupiter platform Engine use  Extensions

**Comment by maciejwalkowiak** [comment]
I don't think this is going to work. `@InjectWireMock` looks only at test classes hierarchy. Perhaps if it would be added as a Spring Bean, you could inject it, but at least at the time when I was designing this library I figured that it creates more noise than needed. Which, may not be true after all.

**Reply by tomakehurst** [reply]
How easy would it be to add a config flag indicating that a Spring bean should also be added?

**Reply by tomasbjerre** [reply]
Relates to https://github.com/wiremock/wiremock-spring-boot/issues/73

Maby we can try to discuss it in that issue.

I'm not sure but I think maby we can make this change while keeping the API as is. By just producing the beans ans change InjectWiremock annotation to just alias Autowire.

Perhaps something to just implement and release, and be prepared to revert if any problems are reported.

**Comment by Arc-E-Tect** [comment]
Not using the integration between WireMock and Spring Boot, but I did manage to get Cucumber, WireMock and Spring Boot to work nicely together for try API-First development. It took some effort, so I decided to write it up: [WireMock, Cucumber, and SpringBoot](https://medium.com/@arc-e-tect/wiremock-cucumber-and-springboot-ae3e107bd3d7)
